crates/node-bindings/src/helpers/js_callable/js_callable.rs

```diff
@@ -0,0 +1,288 @@
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::panic;
+use std::rc::Rc;
+use std::sync::mpsc::channel;
+use std::sync::mpsc::Sender;
+use std::thread::ThreadId;
+
+use super::JsValue;
+use erased_serde::Serialize as ErasedSerialize;
+use napi::bindgen_prelude::Array;
+use napi::bindgen_prelude::FromNapiValue;
+use napi::threadsafe_function::ErrorStrategy;
+use napi::threadsafe_function::ThreadSafeCallContext;
+use napi::threadsafe_function::ThreadsafeFunction;
+use napi::threadsafe_function::ThreadsafeFunctionCallMode;
+use napi::Env;
+use napi::JsFunction;
+use napi::JsObject;
+use napi::JsUnknown;
+use napi::NapiRaw;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+
+thread_local! {
+  /// Storage for napi JavaScript functions on the local thread
+  static LOCAL_FUNCTIONS: (RefCell<usize>, RefCell<HashMap<usize, Rc<JsValue>>>) = Default::default();
```

Comment:
I think everything looks pretty good, except for the unsafe & type-erasure this introduces.

But IMO it's much much simpler to just have 2 forms of delegation:

* One is using multi-threaded function & delegates using the `create_js_thread_safe_method` and `create_async_js_thread_safe_method`, which are combined around 40 lines of code with no unsafe
* The other can use normal delegation - https://github.com/parcel-bundler/parcel/blob/25b7f904fa7511b0a88b7c90252d4d66d51014cf/crates/parcel_filesystem/src/js_delegate_file_system.rs#L44

If we found that there is a need to write an abstraction that works for both cases, we can use it.

Reply:
Also we chatted re. `env` not being safe to store across calls;

The contents of this hashmap are thread-local copies of the functions and env pointer.

My understanding is this needs to be `Ref` because these values may be GCed across calls.

Reply:
> If we found that there is a need to write an abstraction that works for both cases, we can use it.

Perhaps you're right about this. We will probably never call a function from the main JavaScript thread

> My understanding is this needs to be Ref because these values may be GCed across calls.

Good point, I'll double check this

Reply:
Removed local functions